### DISCUSSION TITLE
Design philosophy

### DISCUSSION THREAD

**Discussion by pike2k** [tjko/fanpico#127] [General]
Hi there. 
I wonder why you choose to use 'PWM input' for Tachometer reading and not PIO?
What are the pro's and con's? Maybe you have a special reason to go this route with this project?

**Comment by tjko** [comment]
@pike2k, did you mean reading fan tachometer pulses received or reading fan PWM control signal (from motherboard)?

For reading/decoding the incoming PWM signal, RP2040 PWM hardware is used.  While for reading the tachometer pulses from fans,  interrupts are used (as these pulses come at very "slow" rate).

I think I implemented tachometer reading based on how I had done it on Arduino in the past, not pausing to think if there was better way to do it. 
Using PIO for this sounds like a good idea, especially if could make PIO control the output pins for the multiplexer...

**Reply by pike2k** [reply]
I meant reading tachometer pulses from fans. 
The reason I ask is because I'm working on my own project that (among things) controls a maximum of 4 Noctua fans. I'm gonna have the Pico do many other things and I don't want to slow it down more than necessary with the Tachometer reading. In the past I've noticed that using Interrupts on ESP series chips for Tachometer works poorly if you want to do other things with the CPU.

**Reply by tjko** [reply]
Using interrupts (to "count" pulses) seems to work fine on RP2040, and allows measurement to happen on "background" while CPU does other things...
But using PIO likely would work even better, should be able to have PIO continuously measure pulse length and return latest measurement result when CPU ask for it...

**Reply by pike2k** [reply]
It looks like you use Pio today to generate PWM signal? How many of the 8 statemachines are free on Pico(1) today? Or you mean to switch and use Pico's native PWM for PWM out and attempt to use PIO for fan Tachometer reading?

**Reply by tjko** [reply]
I'm using PWM hardware to generate PWM signals for the fans and read PWM signals from "motherboard". That uses up all 8 PWM blocks on RP2040, 4 blocks needed to output 8 PWM signals and another 4 blocks needed to read 4 signals (as only the "B" channel can be used for PWM input on each block).

PIO is currently used to generate (virtual) tachometer signal that is sent to motherboard. That uses four PIO state machines and  WiFi steals one (if using Pico W), so only left with 3 free PIO state machines.   Probably don't have enough state machines free to use PIO for reading fan tachometer signals (on boards that don't use multiplexer on incoming tachometer signal from fans).

**Reply by pike2k** [reply]
I see. So with a Pico2 you would have 8 free SM's unless W model.
